sdk/python/feast/pyspark/batch_retrieval_job.py

```diff
@@ -0,0 +1,378 @@
+import json
+from typing import Any, Dict, List
+
+from pyspark import SparkFiles
+from pyspark.sql import DataFrame, SparkSession, Window
+from pyspark.sql.functions import col, expr, monotonically_increasing_id, row_number
+
+
+def as_of_join(
+    entity: DataFrame,
+    join_keys: List[str],
+    feature_table: DataFrame,
+    features: List[str],
+    feature_prefix: str = "",
+    max_age: str = None,
+) -> DataFrame:
+    """Perform an as of join between entity and feature table, given a maximum age tolerance.
+    Join conditions:
+    1. Join keys values match.
+    2. Feature event timestamp is the closest match possible to the entity event timestamp,
+       but must not be more recent than the entity event timestamp, and the difference must
+       not be greater than max_age, unless max_age is not specified.
+    3. If more than one feature table rows satisfy condition 1 and 2, feature row with the
+       most recent created timestamp will be chosen.
+
+    Args:
+        entity (DataFrame):
+            Entity dataframe. Must contain the column event_timestamp.
+        join_keys (List[str]):
+            Column names for the join keys.
+        feature_table (DataFrame):
+            Feature table dataframe. Must contain the columns event_timestamp and created_timestamp.
+        features (List[str]):
+            The feature columns which should be present in the result dataframe.
+        feature_prefix (str):
+            Feature column prefix for the result dataframe.
+        max_age (str):
```

Comment:
why it's `str`? what's the format?

Reply:
This is for the Pyspark `interval` expression. For example, 13 hour, 6 day, 60 second, 3 month. I can also change this to integer type instead of string, in which case the max age would be in seconds, similar to the current behaviour. Just thought that using a string might be more convenient for users, since they don't need to convert days / months to seconds, though that would be a breaking change. So should i use seconds and integer instead?